README.md
# Thundercard

![](https://badgen.net/github/checks/Dev-roll/thundercard/main)
![](https://badgen.net/github/release/Dev-roll/thundercard)

Thundercard is a completely new SNS inspired by business cards. It is ideal for the young people, engineers and creators who have a lot of profiles to introduce.

<a href="https://thundercard.gajeroll.com" target="_blank" rel="noreferrer">
  <img width="2560" alt="Thundercard" src="https://user-images.githubusercontent.com/79978827/202828236-809fb30e-9d46-46aa-87da-18882e1f35b1.png">
</a>

## Features

- The world's fastest and most stress-free profile exchange.
- Customize your card with a variety of themes.
- Access the SNS and contacts with a single tap.
- Chat with the people you have exchanged cards with.
- Beautiful human-centered UI.

## Supported Platforms

Thundercard has been built to support multiple platforms. These include:

- Android ([Google Play](https://play.google.com/store/apps/details?id=app.web.thundercard))
- iOS ([App Store](https://apps.apple.com/us/app/thundercard/id6444392874))
- Web ([thundercard-test.web.app](https://thundercard-test.web.app))

## Development

This project uses [Flutter](https://flutter.dev).

Install the following tools, if you don't have them already:

- [Flutter](https://flutter.dev/docs/get-started/install)
- [Firebase CLI](https://firebase.google.com/docs/cli#install_the_firebase_cli)

One can run this app locally for any of these platforms.
To develop locally, clone this repository and run the following commands:

```bash
flutter pub get
flutter run
```

## Contributers

- [@notchcoder](https://github.com/notchcoder)
- [@cardseditor](https://github.com/cardseditor)
- [@keigomichi](https://github.com/keigomichi)


CHANGELOG.md
## 1.0.8 (2022-11-27)

- feat: レスポンシブ対応 (#464) ([75a3782](https://github.com/Dev-roll/thundercard/commit/75a378246bc0094e59cba7411ff8b9c9f8c1a3ab)), closes [#464](https://github.com/Dev-roll/thundercard/pull/464)

## 1.0.7 (2022-11-23)

- fix: SnackBar の不具合を修正 (#458) ([35adc60](https://github.com/Dev-roll/thundercard/commit/35adc60e7f2b115915d9c2131143fa5e56fe14ce)), closes [#458](https://github.com/Dev-roll/thundercard/pull/458)

## 1.0.6 (2022-11-22)

- chore: レスポンシブ対応: upload_image_page.dart (#464) ([a65d8f4](https://github.com/Dev-roll/thundercard/pull/464/commits/a65d8f47953b6185a618fa27f46f54022edbc192)), closes [#464](https://github.com/Dev-roll/thundercard/pull/464)
- 14(1.0.6) 2022/11/22 アカウント登録の不具合を修正

## 1.0.5 (2022-11-20)

- chore: update icon, version (#454) ([45f6451](https://github.com/Dev-roll/thundercard/commit/45f64512cad82761c255f17b60a56d8db6a76920)), closes [#454](https://github.com/Dev-roll/thundercard/pull/454)
- chore: fix (#450) ([7b714d8](https://github.com/Dev-roll/thundercard/commit/7b714d8e2cee431d17bcaf5e2db6f5caf36a7a2c)), closes [#450](https://github.com/Dev-roll/thundercard/pull/450)
- fix: 全てのプラットフォームで通知のページを表示 (#447) ([e057c13](https://github.com/Dev-roll/thundercard/commit/e057c1333d29352fbbe18a1f6034589368928308)), closes [#447](https://github.com/Dev-roll/thundercard/pull/447)

## 1.0.4 (2022-11-18)

- fix: 認証方法追加時の処理を修正 (#441) ([3ad4dff](https://github.com/Dev-roll/thundercard/commit/3ad4dff7039669273fb22fede1c95a81832ffa64)), closes [#441](https://github.com/Dev-roll/thundercard/pull/441)
- fix: debug (#442) ([f1e19dc](https://github.com/Dev-roll/thundercard/commit/f1e19dccf27f849adbc8827adc7baf360a4504ba)), closes [#442](https://github.com/Dev-roll/thundercard/pull/442)

## 1.0.3 (2022-11-17)

- feat: App Store リリース／アカウント機能を更新 (#438) ([c051856](https://github.com/Dev-roll/thundercard/commit/c05185608e83e20dcfad9c3d059329aeeddd3033)), closes [#438](https://github.com/Dev-roll/thundercard/pull/438)

## 1.0.2 (2022-11-15)

- chore: ios release (#437) ([c08703c](https://github.com/Dev-roll/thundercard/commit/c08703c035470adb922d7170e0e6eecc74bb4c22)), closes [#437](https://github.com/Dev-roll/thundercard/pull/437)
- chore: 利用規約を追加 (#435) ([8d19b07](https://github.com/Dev-roll/thundercard/commit/8d19b0728ffa1a7ac961f9fb4a559f9d42498d16)), closes [#435](https://github.com/Dev-roll/thundercard/pull/435)

## 1.0.1 (2022-11-09)

- 交換に関する不具合を修正

## 1.0.0 (2022-11-08)

- Google Play リリース
- Google Play 提出


lib/list.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_speed_dial/flutter_speed_dial.dart';
import 'package:thundercard/widgets/scan_qr_code.dart';
import 'package:thundercard/widgets/custom_skeletons/skeleton_card.dart';

import 'api/colors.dart';
import 'api/current_brightness.dart';
import 'api/current_brightness_reverse.dart';
import 'api/firebase_auth.dart';
import 'widgets/custom_progress_indicator.dart';
import 'widgets/my_card.dart';
import 'card_details.dart';
import 'constants.dart';
import 'search.dart';
import 'upload_image_page.dart';

class List extends StatefulWidget {
  const List({
    Key? key,
  }) : super(key: key);

  @override
  State<List> createState() => _ListState();
}

class _ListState extends State<List> {
  final String? uid = getUid();
  CollectionReference users = FirebaseFirestore.instance.collection('users');
  CollectionReference cards = FirebaseFirestore.instance.collection('cards');
  Map<String, dynamic>? data;
  var isDialOpen = ValueNotifier<bool>(false);
  var customDialRoot = false;
  var buttonSize = const Size(56.0, 56.0);
  var childrenButtonSize = const Size(56.0, 56.0);
  var extend = false;
  var visible = true;
  var rmicons = false;
  var myCardId = '';

  @override
  Widget build(BuildContext context) {
    SystemChrome.setSystemUIOverlayStyle(
      SystemUiOverlayStyle(
        systemNavigationBarColor: alphaBlend(
            Theme.of(context).colorScheme.primary.withOpacity(0.08),
            Theme.of(context).colorScheme.surface),
        statusBarIconBrightness:
            currentBrightnessReverse(Theme.of(context).colorScheme),
        statusBarBrightness: currentBrightness(Theme.of(context).colorScheme),
        statusBarColor: Colors.transparent,
      ),
    );
    return FutureBuilder<DocumentSnapshot>(
      future: users.doc(uid).get(),
      builder:
          (BuildContext context, AsyncSnapshot<DocumentSnapshot> snapshot) {
        if (snapshot.hasError) {
          return const Text('問題が発生しました');
        }

        if (snapshot.hasData && !snapshot.data!.exists) {
          return const Text('ユーザー情報の取得に失敗しました');
        }

        if (snapshot.connectionState == ConnectionState.done) {
          Map<String, dynamic> user =
              snapshot.data!.data() as Map<String, dynamic>;
          return Scaffold(
            // appBar: AppBar(),
            body: SafeArea(
              child: Center(
                child: StreamBuilder<DocumentSnapshot<Object?>>(
                  stream: cards.doc(user['my_cards'][0]).snapshots(),
                  builder: (BuildContext context,
                      AsyncSnapshot<DocumentSnapshot> snapshot) {
                    if (snapshot.hasError) {
                      return const Text('問題が発生しました');
                    }
                    if (snapshot.connectionState == ConnectionState.waiting) {
                      return const CustomProgressIndicator();
                    }
                    dynamic data = snapshot.data;
                    final exchangedCards = data?['exchanged_cards'];
                    final exchangedCardsLength = exchangedCards?.length ?? 0;

                    return Column(
                      children: [
                        ConstrainedBox(
                          constraints: const BoxConstraints(
                            maxWidth: 720,
                          ),
                          child: Container(
                            height: 52,
                            margin: const EdgeInsets.fromLTRB(24, 16, 24, 8),
                            decoration: BoxDecoration(
                              color: Theme.of(context)
                                  .colorScheme
                                  .surfaceVariant
                                  .withOpacity(0.5),
                              borderRadius: BorderRadius.circular(30),
                            ),
                            child: GestureDetector(
                              behavior: HitTestBehavior.opaque,
                              onTap: () {
                                Navigator.of(context).push(
                                  PageRouteBuilder(
                                    pageBuilder: (context, animation,
                                            secondaryAnimation) =>
                                        Search(
                                            exchangedCardIds: exchangedCards),
                                    transitionDuration:
                                        const Duration(seconds: 0),
                                  ),
                                );
                              },
                              child: Row(
                                children: [
                                  Container(
                                    padding: const EdgeInsets.fromLTRB(
                                        20, 12, 0, 12),
                                    child: Icon(
                                      Icons.search_rounded,
                                      color: Theme.of(context)
                                          .colorScheme
                                          .onSurfaceVariant,
                                    ),
                                  ),
                                  const SizedBox(width: 4),
                                  Expanded(
                                    child: Container(
                                      margin: const EdgeInsets.only(right: 16),
                                      child: TextField(
                                        enabled: false,
                                        decoration: const InputDecoration(
                                          hintText: 'カードを検索',
                                          filled: true,
                                          fillColor: Colors.transparent,
                                          disabledBorder: UnderlineInputBorder(
                                            borderSide: BorderSide(
                                              color: Colors.transparent,
                                              width: 0,
                                            ),
                                          ),
                                        ),
                                        onChanged: ((value) {}),
                                      ),
                                    ),
                                  ),
                                ],
                              ),
                            ),
                          ),
                        ),
                        (exchangedCardsLength != 0)
                            ? Expanded(
                                child: ListView.builder(
                                  // shrinkWrap: true,
                                  // physics:
                                  //     const NeverScrollableScrollPhysics(),
                                  itemCount: exchangedCards.length + 2,
                                  itemBuilder: (context, index) {
                                    if (index == 0) {
                                      return const SizedBox(height: 16);
                                    }
                                    if (index == exchangedCards.length + 1) {
                                      return const SizedBox(height: 80);
                                    }
                                    return StreamBuilder<
                                        DocumentSnapshot<Object?>>(
                                      stream: cards
                                          .doc(exchangedCards[index - 1])
                                          .snapshots(),
                                      builder: (BuildContext context,
                                          AsyncSnapshot<DocumentSnapshot>
                                              snapshot) {
                                        if (snapshot.hasError) {
                                          return const Text('問題が発生しました');
                                        }
                                        if (snapshot.connectionState ==
                                            ConnectionState.waiting) {
                                          return const SkeletonCard();
                                        }
                                        dynamic card = snapshot.data;
                                        if (!snapshot.hasData) {
                                          return const Text('カードの情報の取得に失敗しました');
                                        }
                                        return Column(
                                          children: [
                                            GestureDetector(
                                              behavior: HitTestBehavior.opaque,
                                              onTap: () {
                                                Navigator.of(context)
                                                    .push(MaterialPageRoute(
                                                  builder: (context) =>
                                                      CardDetails(
                                                    cardId: exchangedCards[
                                                        index - 1],
                                                    card: card,
                                                  ),
                                                ));
                                              },
                                              child: ConstrainedBox(
                                                constraints:
                                                    const BoxConstraints(
                                                  maxHeight: 400,
                                                ),
                                                child: FittedBox(
                                                  child: MyCard(
                                                    cardId: exchangedCards[
                                                        index - 1],
                                                    cardType: CardType.normal,
                                                  ),
                                                ),
                                              ),
                                            ),
                                            const SizedBox(height: 24),
                                          ],
                                        );
                                      },
                                    );
                                  },
                                ),
                              )
                            : Expanded(
                                child: Column(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    Icon(
                                      Icons.priority_high_rounded,
                                      size: 120,
                                      color: Theme.of(context)
                                          .colorScheme
                                          .onBackground
                                          .withOpacity(0.3),
                                    ),
                                    const SizedBox(height: 20),
                                    Text(
                                      'まだカードがありません',
                                      style: TextStyle(
                                          color: Theme.of(context)
                                              .colorScheme
                                              .onSurfaceVariant),
                                    ),
                                  ],
                                ),
                              ),
                      ],
                    );
                  },
                ),
              ),
            ),
            floatingActionButtonLocation: FloatingActionButtonLocation.endFloat,
            floatingActionButton: SpeedDial(
              // animatedIcon: AnimatedIcons.menu_close,
              animatedIconTheme: const IconThemeData(size: 24.0),
              // / This is ignored if animatedIcon is non null
              // child: Text('open'),
              // activeChild: Text('close'),
              icon: Icons.add_rounded,
              activeIcon: Icons.close_rounded,
              foregroundColor:
                  Theme.of(context).colorScheme.onSecondaryContainer,
              backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
              spacing: 16,
              openCloseDial: isDialOpen,
              activeBackgroundColor:
                  Theme.of(context).colorScheme.secondaryContainer,
              childPadding: const EdgeInsets.all(0),
              spaceBetweenChildren: 0,
              dialRoot: customDialRoot
                  ? (ctx, open, toggleChildren) {
                      return ElevatedButton(
                        onPressed: toggleChildren,
                        style: ElevatedButton.styleFrom(
                          foregroundColor:
                              Theme.of(context).colorScheme.secondaryContainer,
                          padding: const EdgeInsets.symmetric(
                              horizontal: 22, vertical: 18),
                        ),
                        child: const Text(
                          'Custom Dial Root',
                          style: TextStyle(fontSize: 17),
                        ),
                      );
                    }
                  : null,
              buttonSize:
                  buttonSize, // it's the SpeedDial size which defaults to 56 itself
              iconTheme: const IconThemeData(size: 24),
              label: extend
                  ? const Text('Open')
                  : null, // The label of the main button.
              /// The active label of the main button, Defaults to label if not specified.
              activeLabel: extend ? const Text('Close') : null,

              /// Transition Builder between label and activeLabel, defaults to FadeTransition.
              // labelTransitionBuilder: (widget, animation) => ScaleTransition(scale: animation,child: widget),
              /// The below button size defaults to 56 itself, its the SpeedDial childrens size
              childrenButtonSize: childrenButtonSize,
              visible: visible,
              direction: SpeedDialDirection.up,
              switchLabelPosition: false,

              /// If true user is forced to close dial manually
              closeManually: false,

              /// If false, backgroundOverlay will not be rendered.
              renderOverlay: true,
              // overlayColor: Colors.black,
              overlayOpacity: 0.9,
              // onOpen: () => debugPrint('OPENING DIAL'),
              // onClose: () => debugPrint('DIAL CLOSED'),
              useRotationAnimation: true,
              tooltip: '',
              heroTag: 'speed-dial-hero-tag',
              // foregroundColor: Colors.black,
              // backgroundColor: Colors.white,
              // activeForegroundColor: Colors.red,
              // activeBackgroundColor: Colors.blue,
              elevation: 8.0,
              animationCurve: Curves.easeInOut,
              isOpenOnStart: false,
              animationDuration: const Duration(milliseconds: 200),
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(16),
              ),
              // childMargin: EdgeInsets.symmetric(horizontal: 10, vertical: 5),
              childMargin: const EdgeInsets.fromLTRB(0, 0, 8, 0),
              children: [
                SpeedDialChild(
                  child: !rmicons
                      ? Container(
                          padding: const EdgeInsets.all(8),
                          margin: const EdgeInsets.all(8),
                          decoration: BoxDecoration(
                            color: Theme.of(context).colorScheme.primary,
                            borderRadius: BorderRadius.circular(12),
                            boxShadow: const [
                              BoxShadow(
                                color: Color(0xaa000000),
                                blurRadius: 8,
                                offset: Offset(0, 4),
                                spreadRadius: -2,
                              )
                            ],
                          ),
                          child: Icon(
                            Icons.add_a_photo_rounded,
                            color: Theme.of(context).colorScheme.onPrimary,
                          ),
                        )
                      : null,
                  backgroundColor: Colors.transparent,
                  foregroundColor:
                      Theme.of(context).colorScheme.onSecondaryContainer,
                  label: '画像をもとに追加',
                  onTap: () {
                    Navigator.of(context).push(MaterialPageRoute(
                      builder: (context) => UploadImagePage(
                        cardId: user['my_cards'][0],
                      ),
                      fullscreenDialog: true,
                    ));
                  },
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(12),
                  ),
                  labelBackgroundColor: Colors.transparent,
                  labelShadow: [],
                  elevation: 0,
                ),
                SpeedDialChild(
                  child: !rmicons
                      ? Container(
                          padding: const EdgeInsets.all(8),
                          margin: const EdgeInsets.all(8),
                          decoration: BoxDecoration(
                            color: Theme.of(context).colorScheme.primary,
                            borderRadius: BorderRadius.circular(12),
                            boxShadow: const [
                              BoxShadow(
                                color: Color(0xaa000000),
                                blurRadius: 8,
                                offset: Offset(0, 4),
                                spreadRadius: -2,
                              )
                            ],
                          ),
                          child: Icon(
                            Icons.qr_code_scanner_rounded,
                            color: Theme.of(context).colorScheme.onPrimary,
                          ),
                        )
                      : null,
                  backgroundColor: Colors.transparent,
                  foregroundColor:
                      Theme.of(context).colorScheme.onSecondaryContainer,
                  label: 'カードを交換',
                  onTap: () {
                    Navigator.of(context).push(MaterialPageRoute(
                      builder: (context) => Theme(
                        data: ThemeData(
                          colorSchemeSeed:
                              Theme.of(context).colorScheme.secondaryContainer,
                          brightness: Brightness.dark,
                          useMaterial3: true,
                        ),
                        child: ScanQrCode(myCardId: myCardId),
                      ),
                    ));
                  },
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(12),
                  ),
                  labelBackgroundColor: Colors.transparent,
                  labelShadow: [],
                  elevation: 0,
                ),
              ],
            ), // floatingActionButton: FloatingActionButton.extended(
          );
        }
        return const Scaffold(
          body: Center(child: CustomProgressIndicator()),
        );
      },
    );
  }
}
